lib/spack/spack/fetch_strategy.py

```diff
@@ -768,173 +896,216 @@ def __init__(self, **kwargs):
+
+    @staticmethod
+    @memoized
+    def _get_git_version(git):
+        vstring = git('--version', output=str).lstrip('git version ')
         return Version(vstring)
 
-    @property
+    @property                                               # type: ignore[misc]
+    @memoized
     def git(self):
-        if not self._git:
-            self._git = which('git', required=True)
+        git = which('git', required=True)
 
-            # Disable advice for a quieter fetch
-            # https://github.com/git/git/blob/master/Documentation/RelNotes/1.7.2.txt
-            if self.git_version >= Version('1.7.2'):
-                self._git.add_default_arg('-c')
-                self._git.add_default_arg('advice.detachedHead=false')
+        # Disable advice for a quieter fetch
+        # https://github.com/git/git/blob/master/Documentation/RelNotes/1.7.2.txt
+        if self._get_git_version(git) >= Version('1.7.2'):
+            git.add_default_arg('-c')
+            git.add_default_arg('advice.detachedHead=false')
 
-            # If the user asked for insecure fetching, make that work
-            # with git as well.
-            if not spack.config.get('config:verify_ssl'):
-                self._git.add_default_env('GIT_SSL_NO_VERIFY', 'true')
+        # If the user asked for insecure fetching, make that work
+        # with git as well.
+        if not spack.config.get('config:verify_ssl'):
+            git.add_default_env('GIT_SSL_NO_VERIFY', 'true')
 
-        return self._git
+        return git
 
     @property
     def cachable(self):
-        return self.cache_enabled and bool(self.commit or self.tag)
+        return (self.cache_enabled and
+                not self.ref.always_check_for_updates_but_accept_airgapped())
 
     def source_id(self):
-        return self.commit or self.tag
+        return self.ref.refspec()
 
     def mirror_id(self):
-        repo_ref = self.commit or self.tag or self.branch
-        if repo_ref:
-            repo_path = url_util.parse(self.url).path
-            result = os.path.sep.join(['git', repo_path, repo_ref])
-            return result
+        repo_path = url_util.parse(self.url).path
+        return os.path.sep.join(['git', repo_path, self.ref.refspec()])
 
     def _repo_info(self):
-        args = ''
+        return '{0} {1}'.format(self.url, self.ref.repo_info_for_reference())
+
+    @memoized
+    def _cache_path(self):
+        # type: () -> str
+        cache_path = spack.caches.fetch_cache.persistent_cache_dir_for(self)
+        if self._initialize_git_repo(cache_path):
+            tty.debug('Initialized git repository: {0}'.format(self._repo_info()))
+        else:
+            tty.debug('Already fetched {0}'.format(cache_path))
+        return cache_path
+
+    def _dormant_branch_name(self, cache_path):
+        # type: (str) -> str
+        if self.branch == self._special_dormant_branch_name:
+            raise FetcherConflict(dedent("""\
```

Comment:
We should probably have a deep-in-the-weeds configuration option that a user can specify to change the dormant branch name, and this error message should inform the user how to do this.

We should also put the word "spack" into the dormant branch name, so that conflicts are (even) less likely.

Reply:
That sounds great! Will do.

Reply:
Generating a UUID allows us to avoid this!